elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java

```diff
@@ -0,0 +1,169 @@
+    });
+  }
+
+  private Set<ReferenceExpression> extractFields(Expression expr) {
+    Set<ReferenceExpression> fields = new HashSet<>();
+    expr.accept(new ExpressionNodeVisitor<Object, Set<ReferenceExpression>>() {
+      @Override
+      public Object visitReference(ReferenceExpression node, Set<ReferenceExpression> context) {
+        context.add(node);
+        return null;
+      }
+    }, fields);
+    return fields;
+  }
+
+  private ElasticsearchExprValueFactory buildValueFactory(Set<ReferenceExpression> fields) {
+    Map<String, ExprType> typeEnv = fields.stream()
+                                          .collect(toMap(
+                                              ReferenceExpression::getAttr,
+                                              ReferenceExpression::type));
+    return new ElasticsearchExprValueFactory(typeEnv);
+  }
+
+  private Map<String, ExprValue> buildValueEnv(Set<ReferenceExpression> fields,
+                                               ElasticsearchExprValueFactory valueFactory) {
+    Map<String, ExprValue> valueEnv = new HashMap<>();
+    for (ReferenceExpression field : fields) {
+      String fieldName = field.getAttr();
+      ExprValue exprValue = valueFactory.construct(fieldName, getDocValue(field));
+      valueEnv.put(fieldName, exprValue);
+    }
+    return valueEnv;
+  }
+
+  private Object getDocValue(ReferenceExpression field) {
+    String fieldName = getDocValueName(field);
+    ScriptDocValues<?> docValue = getDoc().get(fieldName);
+    if (docValue == null || docValue.isEmpty()) {
+      return null;
+    }
+
+    Object value = docValue.get(0);
+    if (value instanceof ChronoZonedDateTime) {
+      return ((ChronoZonedDateTime<?>) value).toInstant();
+    }
+    return castNumberToFieldType(value, field.type());
+  }
+
+  /**
+   * Text field doesn't have doc value (exception thrown even when you call "get")
+   * Limitation: assume inner field name is always "keyword".
+   */
+  private String getDocValueName(ReferenceExpression field) {
+    String fieldName = field.getAttr();
+    if (field.type() == ES_TEXT_KEYWORD) {
+      fieldName += ".keyword";
+    }
+    return fieldName;
+  }
+
+  /**
+   * DocValue only support long and double so cast to integer and float if needed.
+   * The doc value must be Long and Double for expr type Long/Integer and Double/Float respectively.
+   * Otherwise there must be bugs in our engine that causes the mismatch.
+   */
+  private Object castNumberToFieldType(Object value, ExprType type) {
+    if (type == INTEGER) {
+      return ((Long) value).intValue();
+    } else if (type == FLOAT) {
+      return ((Double) value).floatValue();
+    } else {
+      return value;
+    }
+  }
+
+  private ExprValue evaluateExpression(Map<String, ExprValue> valueEnv) {
+    ExprTupleValue tupleValue = ExprTupleValue.fromExprValueMap(valueEnv);
+    ExprValue result = expression.valueOf(tupleValue.bindingTuples());
+
+    if (result.isNull()) {
```

Comment:
Missing handling?

Reply:
Since no missing concept in JAVA, will add a `MISSING` constant in value factory. And will add test case for `null` field.

Reply:
I added missing value check here. But I found that there seems no way to differentiate null and missing in doc value. I tried the following index. Both `null_field` and `missing_field` in 2nd doc gives a script doc value object with empty list inside. Will confirm and come back to this later.

```
PUT null_and_missing/_doc/1
{
  "null_field": "abc",
  "missing_field": 456
}

PUT null_and_missing/_doc/2
{
  "null_field": null
}
```